### PR TITLE
sql: Implement ALL ROLES for ALTER DEFAULT PRIVs

### LINKED ISSUE #19647: SQL: Support default privileges
### Feature request

PostgreSQL allows setting default privileges on objects. A default privilege controls what privileges an object is assigned when it is first created.

See:
- https://www.postgresql.org/docs/15/sql-alterdefaultprivileges.html
- https://www.postgresql.org/docs/15/ddl-priv.html
- https://www.postgresql.org/docs/15/catalog-pg-default-acl.html

### PR DESCRIPTION
**Pull request by jkosh44**
This commit allows users to alter the default privileges of objects
created by all existing and future roles.

Resolves #19647

### Motivation
This PR adds a known-desirable feature.

### Tips for Reviewer
This PR builds off of https://github.com/MaterializeInc/materialize/pull/19913, only the most recent commit is new.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow admins to alter default privileges for all roles current and future.

### CODE REVIEW

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -3149,9 +3170,16 @@ pub struct AlterDefaultPrivilegesStatement<T: AstInfo> {
...
-            f.write_node(&display::comma_separated(target_roles));
+        match &self.target_roles {
+            TargetRoleSpecification::Roles(_) => {
+                f.write_str(" FOR ROLE ");
+                f.write_node(&self.target_roles);
+            }
+            TargetRoleSpecification::CurrentRole => {}
+            TargetRoleSpecification::AllRoles => {
+                f.write_str(" FOR ");
+                f.write_node(&self.target_roles);
```
Comment (def-) [inline src/sql-parser/src/ast/defs/statement.rs:3181]: Could be tested

File: src/sql-parser/src/ast/defs/statement.rs [hunk 2]
```diff
@@ -3048,6 +3048,27 @@ impl<T: AstInfo> AstDisplay for RevokePrivilegesStatement<T> {
...
+    /// All current and future roles.
+    AllRoles,
+}
+
+impl<T: AstInfo> AstDisplay for TargetRoleSpecification<T> {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        match self {
+            TargetRoleSpecification::Roles(roles) => f.write_node(&display::comma_separated(roles)),
+            TargetRoleSpecification::CurrentRole => {}
+            TargetRoleSpecification::AllRoles => f.write_str("ALL ROLES"),
```
Comment (def-) [inline src/sql-parser/src/ast/defs/statement.rs:3066]: Could be tested

**Review by def- — COMMENTED** [review]
Coverage mostly looks fine: https://buildkite.com/materialize/coverage/builds/139#0188bb8c-6f59-4d4b-acc9-836fea6d49af

File: src/adapter/src/coord/sequencer/inner.rs [hunk 3]
```diff
@@ -3944,7 +3944,7 @@ impl Coordinator {
         };
         for privilege_object in &privilege_objects {
             self.catalog()
-                .ensure_not_reserved_role(&privilege_object.role_id)?;
```
Comment (ParkMyCar) [inline src/adapter/src/coord/sequencer/inner.rs]: For my own knowledge, is the system role the only "reserved" role?

**Review by ParkMyCar — APPROVED** [review]
Only reviewed the most recent commit, but this looks good!

File: src/sql-parser/src/ast/defs/statement.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql-parser/src/ast/defs/statement.rs:3181]: Is it possible that the coverage report is inaccurate here? I'm almost positive that it's getting tested here: https://github.com/MaterializeInc/materialize/pull/19939/commits/82268626e77ff18f89ee346bd77546a7724da961#diff-93af953bf24ae7fafa2260b763b0f4066e9ae6d1645efb6c6492ebc54925523cR1120-R1126

File: src/adapter/src/coord/sequencer/inner.rs (hunk 3 above)
Comment (jkosh44) [reply src/adapter/src/coord/sequencer/inner.rs]: Any role with a system ID (`mz_system` and `mz_introspection`) and `PUBLIC` are reserved roles. We also have some prefixes that we consider reserved when creating a new role, however that's not an issue for already existing roles.

File: src/adapter/src/rbac.rs [hunk 4]
```diff
@@ -225,6 +225,20 @@ pub fn check_plan(
     role_memberships.insert(*current_role_id, role_membership);
     check_object_privileges(catalog, required_privileges, role_memberships)?;
 
+    if let Plan::AlterDefaultPrivileges(AlterDefaultPrivilegesPlan {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/rbac.rs:233]: Done

File: src/sql-parser/src/ast/defs/statement.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql-parser/src/ast/defs/statement.rs:3181]: The test definitely starts failing if I modify these lines, so I'm going to assume that it's some inaccuracy with the coverage metrics.

**Comment by jkosh44** [comment]
TFTR, I'll merge this after merging #19913

File: src/sql-parser/src/ast/defs/statement.rs (hunk 1 above)
Comment (def-) [reply src/sql-parser/src/ast/defs/statement.rs:3181]: Alright, my bad. This seems to have run into one of the coverage bugs.

**Pull request merged**